### PR TITLE
[BOB-17] Implement IPkbCreator

### PR DESCRIPTION
**Pull request by pzajaczkowski**

### CODE REVIEW

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    /// <summary>
+    /// Stores modifies relation between container statement(while, if) and variable list
+    /// </summary>
+    /// <remarks>[statement(while, if) line number,  list of variable names].</remarks>
+    private readonly Dictionary<int, List<string>> _containerModifiesDictionary = new();
+
+    public void SetModifies(TreeNode modifier, TreeNode variable)
+    {
+        // Check types
+        if (modifier.Type.IsStatement() == false || variable.Type != EntityType.Variable)
```
[the remark this answers is not in the document]
Comment (maksimowiczm) [reply src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs:28]: Why there is method `IsStatement` but not `IsVariable`

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            logger.LogError(
+                "Statement modifies relation can only be established between statement and variable node. ({modifier} => {variable})",
+                modifier, variable);
+
+            throw new ArgumentException(
+                $"At least one of provided nodes type: {modifier.Type}, {variable.Type} is different than any of required: {EntityType.Statement} types or {EntityType.Variable}.");
```
Comment (maksimowiczm) [inline src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs:35]: Is there any reason why we can't tell which one has invalid type and pass it as a message?

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/FollowsRelation.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /// <summary>
+    /// Stores follows relation between two statements using their line numbers.
+    /// </summary>
+    /// <remarks>[following node line number, preceding node line number].</remarks>
+    private readonly Dictionary<int, int> _followsDictionary = new();
+
+    public void SetFollows(TreeNode node, TreeNode parent)
+    {
+        if (node.Type.IsStatement() == false || parent.Type.IsStatement() == false)
```
Comment (maksimowiczm) [inline src/Pkb/Parsobober.Pkb.Relations/Implementations/FollowsRelation.cs:22]: don't do `smth == false`

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/FollowsRelation.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
+using System.Xml;
```
Comment (maksimowiczm) [inline src/Pkb/Parsobober.Pkb.Relations/Implementations/FollowsRelation.cs:1]: unused using

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/ParentRelation.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+public class ParentRelation(
+    ILogger<ParentRelation> logger,
+    IProgramContextAccessor programContext
+) : IParentCreator
+{
+    /// <summary>
+    /// Stores parent relation between two statements using their line numbers.
+    /// </summary>
+    /// <remarks>[parent node line number, child node line number].</remarks>
+    private readonly Dictionary<int, int> _parentsDictionary = new();
```
Comment (maksimowiczm) [inline src/Pkb/Parsobober.Pkb.Relations/Implementations/ParentRelation.cs:17]: Parent can only have one child.

**Review by maksimowiczm — CHANGES_REQUESTED** [review]

File: src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
+using Microsoft.Extensions.Logging;
+using Parsobober.Pkb.Ast;
+using Parsobober.Pkb.Relations.Abstractions.Accessors;
+using Parsobober.Pkb.Relations.Abstractions.Creators;
+
+namespace Parsobober.Pkb.Relations.Implementations;
+
+public class ModifiesRelation(
+    ILogger<ParentRelation> logger,
+    IProgramContextAccessor programContext
```
[the remark this answers is not in the document]
Comment (pzajaczkowski) [reply src/Pkb/Parsobober.Pkb.Relations/Implementations/ModifiesRelation.cs:10]: unobvious hint for klaudia to use it

**Review by maksimowiczm — APPROVED** [review]

**Review by bciereszynski — APPROVED** [review]

**Pull request merged**